clip-path update (#39165)
mdn/content: files/en-us/web/css/clip-path/index.md

## Examples

### Shapes and geometry boxes

In this example, a triangle is created by defining a `polygon()` as the clip path on an element with a solid colored background with large {{cssxref("border")}}. The second element has a `<>:

#### HTML

```html live-sample___shapes1 live-sample___shapes2 live-sample___shapes3
<div></div>
<div></div>
```

#### CSS

```css hidden live-sample___shapes1 live-sample___shapes2 live-sample___shapes3
body {
  display: flex;
  gap: 20px;
  flex-flow: row wrap;
}
```

```css live-sample___shapes1 live-sample___shapes2 live-sample___shapes3
div {
  height: 200px;
  width: 200px;
  box-sizing: border-box;
  background-color: rebeccapurple;
  border: 20px solid magenta;

  clip-path: polygon(50% 0, 100% 100%, 0 100%);
}

div:last-of-type {
  clip-path: content-box polygon(50% 0, 100% 100%, 0 100%);
}
```

#### Results

{{EmbedLiveSample("shapes1", "", "230")}}

> For the first triangle, we didn't specify a reference box, defaulting to `border-box` being used as the reference box, with the 0% and 100% being the outside edge of the border. In the second example, we set the `<geometry-box>` to `content-box`, meaning the reference box for the basic shape is the outer edge of the content area, which is inside the padding box. In our padding-less example, this is the inner edge of the border.

Reviewer: ```suggestion
For the first triangle, we didn't specify a reference box; it therefore defaults to `border-box`, with the 0% and 100% positions located on the outside edge of the border. In the second example, we set the `<geometry-box>` to `content-box`, meaning the reference box for the basic shape is the outer edge of the content area, which is inside the padding box. Because our example has no `padding`, this is the inner edge of the border.
```

> ### Shape() versus path() functions

Reviewer: ```suggestion
### `shape()` versus `path()` functions
```

> Expanding on the previous example, we create the same shape with different `<basic-shape>` values, demonstrate how the {{cssxref("basic-shape/shape","shape()")}}`and {{cssxref("basic-shape/path","path()")}} functions can also be used to create clipping paths, with`shape()` being a more flexible solution.

Reviewer: ```suggestion
Expanding on the previous example, we create the same triangle with different `<basic-shape>` values, demonstrating how the {{cssxref("basic-shape/shape", "shape()")}} and {{cssxref("basic-shape/path", "path()")}} functions can also be used to create clipping paths, with `shape()` being a more flexible solution.
```

> We use `path()` to define the first element's clipping path, and `shape()` for the second to create a triangle, both using the default `border-box` as their references box:

Reviewer: ```suggestion
We use `path()` to define the first element's clipping path, and `shape()` for the second, both using the default `border-box` as their reference box:
```

```css live-sample___shapes2 live-sample___shapes3
div {
  clip-path: path("M100 0 L200 200 L0 200 Z");
}

div:last-of-type {
  clip-path: shape(from 50% 0, line to 100% 100%, line to 0 100%, close);
}
```

{{EmbedLiveSample("shapes2", "", "230")}}

Because the `shape()` function allows using {{cssxref("percentage")}} values (and {{cssxref("--*", "custom properties")}} too), it is more robust.

> Expanding on the previous result, the only change we make here is an increase the size of the underlying element:

Reviewer: ```suggestion
We'll demonstrate this by increasing the size of the underlying element:
```

```css live-sample___shapes3
div {
  width: 250px;
  height: 250px;
}
```

> 

Reviewer: Add the following sentence before the live example, to introduce it:

> As a result, the path defined with the `shape()` function grows with the element, whereas the `path()` version does not:

{{EmbedLiveSample("shapes3", "", "280")}}

> The path defined with the `shape()` function grew with the element. You'll note by the visibility or at least partial visibility of the four border sides in the clip path example defined by the `shape()` function; by using percentages, the path grew with the element. In the `path()` version, the top and left borders being partially visible while the right and bottom borders are clipped out.

Reviewer: ```suggestion
Note the visibility or at least partial visibility of the four border sides in the clip path example defined by the `shape()` function: the percentage values have allowed the path to grow with the element. In the `path()` version, the element grew but not the shape — as a result, the top and left borders are partially visible while the right and bottom borders are clipped out.
```

Reviewer: Does this work? I want to spell out a bit more explicitly ehy this demonstrates the robustness of the `shape()` function.

### SVG as clip source

In this example, we define SVG {{svgElement("clipPath")}} elements to use as a `clip-path` source.

#### HTML

> We include two {{htmlElement("div")}} elements that we will style along with an SVG element with two `<clipPath>` elements; one containing four {{svgElement("rect")}} elements that together define window panes, leaving a cross of blank space in the middle, and the other containing two crossing `<rect>` elements.

Reviewer: ```suggestion
We include two {{htmlElement("div")}} elements and an `<svg> element containing two `<clipPath>` elements. One `<clipPath>` contains four {{svgElement("rect")}} elements that together define window panes, leaving a cross of blank space in the middle, and the other contains two crossing `<rect>` elements.
```

> We include two {{htmlElement("div")}} elements and an `<svg> element containing two `<clipPath>` elements. One `<clipPath>` contains four {{svgElement("rect")}} elements that together define window panes, leaving a cross of blank space in the middle, and the other contains two crossing `<rect>` elements.

Author: ```suggestion
We include two {{htmlElement("div")}} elements and an `<svg>` element containing two `<clipPath>` elements. One `<clipPath>` contains four {{svgElement("rect")}} elements that together define window panes, leaving a cross of blank space in the middle, and the other contains two crossing `<rect>` elements.
```

```html
<svg height="0" width="0">
  <defs>
    <clipPath id="window">
      <rect y="0" x="0" width="80" height="80" />
      <rect y="0" x="120" width="80" height="80" />
      <rect y="120" x="0" width="80" height="80" />
      <rect y="120" x="120" width="80" height="80" />
    </clipPath>
    <clipPath id="cross">
      <rect y="0" x="80" width="40" height="200" />
      <rect y="80" x="0" width="200" height="40" />
    </clipPath>
  </defs>
</svg>

<div class="window">Window</div>
<div class="cross">Cross</div>
```

#### CSS

> We use flexible layout to allow our elements to sit side-by-side with a gap between if they fit. We define a {{cssxref("gradient/conic-gradient", "conic-gradient()")}} background image on both `<div>` elements, providing an interesting visual to clip, along with a border.

Reviewer: ```suggestion
We use [flexbox](/en-US/docs/Web/CSS/CSS_flexible_box_layout) to allow our elements to sit side-by-side with a gap between them, if there is space available. We define a {{cssxref("gradient/conic-gradient", "conic-gradient()")}} background image on both `<div>` elements, providing an interesting visual to clip, along with a {{cssxref("border")}}.
```

```css
body {
  display: flex;
  gap: 20px;
  flex-flow: row wrap;
  font: 2em sans-serif;
}

div {
  width: 200px;
  height: 200px;
  background-image: conic-gradient(
    at center,
    rebeccapurple,
    green,
    lightblue,
    rebeccapurple
  );

  border: 5px solid;
  box-sizing: border-box;
}
```

> We then set the `id` of the `<clipPath>` as the `<clip-source>`. We included a {{cssxref("align-content")}} on the `cross` example, centering the text vertically, as otherwise the text would be clipped, as is happening in the `window` example.

Reviewer: ```suggestion
We then set the `id` of the `<clipPath>` as the `<clip-source>`. We center the text in the `cross` example vertically using {{cssxref("align-content")}}, as otherwise the text would be clipped, as is happening in the `window` example.
```

```css
.window {
  clip-path: url(#window);
}

.cross {
  clip-path: url(#cross);
  align-content: center;
}
```

#### Results

{{EmbedLiveSample("SVG as clip source", "", "230")}}

> The element, including the border and text, is clipped, with only the parts overlapping the `<clipPath>` being drawn to the page.

Reviewer: ```suggestion
The elements, including their border and text, are clipped, with only the parts overlapping the `<clipPath>` elements being drawn to the page.
```

### The various value types

This example demonstrates the various values of the `clip-path` property clipping an HTML {{htmlelement("img")}}.

#### HTML

> The HTML includes an image and a `<clipPath>` in the shape of a star. We also included a {{htmlelement("select")}} element which provides various optional values for the `clip-path` property.

Reviewer: ```suggestion
The HTML includes an `<img>` that will be clipped, a star-shaped `<clipPath>`, and a {{htmlelement("select")}} element to choose a `clip-path` property value from.
```

```html
<img id="clipped" src="mdn.svg" alt="MDN logo" />
<svg height="0" width="0">
  <defs>
    <clipPath id="star">
      <path d="M50,0 21,90 98,35 2,35 79,90z">
    </clipPath>
  </defs>
</svg>

<select id="clipPath">
  <option value="none">none</option>
  <option value="circle(100px at 110px 100px)">circle</option>
  <option value="url(#star)" selected>star</option>
  <option value="inset(20px round 20px)">inset</option>
  <option value="rect(20px 150px 200px 20px round 10%)">rect</option>
  <option value="xywh(0 20% 90% 67% round 0 0 5% 5px)">xywh</option>
  <option value="path('M 0 200 L 0,110 A 110,90 0,0,1 240,100 L 200 340 z')">
    path
  </option>
</select>
```

```html hidden
<pre id="log"></pre>
```

#### CSS

The initial rendering includes the star as the `clip-path` source.

```css
#clipped {
  margin-bottom: 20px;
  clip-path: url(#star);
}
```

#### JavaScript

> When you select a new option, an event handler updates the value of the `clip-path`.

Reviewer: ```suggestion
When you select a new option from the `<select>` menu, an event handler updates the value of the `clip-path` set on the `<img>`.
```

```js
const clipPathSelect = document.getElementById("clipPath");
clipPathSelect.addEventListener("change", (evt) => {

>   const x = evt.target.value;

Reviewer: Not a blocker, but maybe use something more intuitive like `selectedPath` rather than `x`?

Author: used `path`

  document.getElementById("clipped").style.clipPath = x;
  log(`clip-path: ${x};`);
});
```

```js hidden
function log(text) {
  const logElement = document.querySelector("#log");
  logElement.innerText = `${text}`;
  logElement.scrollTop = logElement.scrollHeight;
}
```

#### Result

> {{EmbedLiveSample("Complete_example", 230, 280)}}

Reviewer: ```suggestion
{{EmbedLiveSample("Complete_example", 230, 300)}}
```

(it was showing a vertical scroll bar all the time on Fx, and when the `none` value was selected on Chrome/Edge. 300 fixes these issues)

> Select different options to change the `clip-path` value. To clip a background to text, use the {{cssxref("background-clip")}} property.

Reviewer: The second sentence comes across as weird, as setting `background-clip` isn't an option available in this demo. Maybe include this info somewhere else?

Expanding on the previous example, we create the same triangle with different `<basic-shape>` values, demonstrating how the {{cssxref("basic-shape/shape", "shape()")}} and {{cssxref("basic-shape/path", "path()")}} functions can also be used to create clipping paths, with `shape()` being a more flexible solution.

We use `path()` to define the first element's clipping path, and `shape()` for the second, both using the default `border-box` as their reference box:

```css live-sample___shapes2 live-sample___shapes3
div {
  clip-path: path("M100 0 L200 200 L0 200 Z");
}

div:last-of-type {
  clip-path: shape(from 50% 0, line to 100% 100%, line to 0 100%, close);
}
```

As a result, the path defined with the shape() function grows with the element, whereas the path() version does not:

{{EmbedLiveSample("shapes2", "", "230")}}

Because the `shape()` function allows using {{cssxref("percentage")}} values (and {{cssxref("--*", "custom properties")}} too), it is more robust.

We'll demonstrate this by increasing the size of the underlying element:

```css live-sample___shapes3
div {
  width: 250px;
  height: 250px;
}
```

{{EmbedLiveSample("shapes3", "", "280")}}

> The visibility, or at least partial visibility, of the four border sides in the clip path example defined by the `shape()` function exists because the percentage values have allowed the path to grow with the element. In the `path()` version, the element grew but not the shape. As a result, the top and left borders are partially visible while the right and bottom borders are clipped out.

Reviewer: ```suggestion
The visibility, or at least partial visibility, of the four border sides in the clip path example defined by the `shape()` function is due to the percentage values allowing the path to grow with the element. In the `path()` version, the element grew, but not the shape. As a result, the top and left borders are partially visible while the right and bottom borders are clipped out.
```
